test/parachain/tests/integration_test.rs

```diff
@@ -0,0 +1,411 @@
+		async_std::task::block_on(async { Chain::block_hash(&mut client, None).await.unwrap() })
+			.unwrap();
+	let current_block = async_std::task::block_on(async {
+		Chain::header(&mut client, current_block_hash)
+			.await
+			.unwrap()
+	})
+	.unwrap()
+	.number
+	.saturated_into::<u64>();
+
+	let genesis_block =
+		async_std::task::block_on(async { Chain::block_hash(&mut client, 0).await.unwrap() })
+			.unwrap();
+
+	// create and sign transaction
+	let wasm =
+		fs::read(target_dir().join(
+			"wbuild/cumulus-test-parachain-runtime/cumulus_test_parachain_runtime.compact.wasm",
+		))
+		.unwrap();
+	let call = pallet_sudo::Call::sudo(Box::new(
+		registrar::Call::<Runtime>::register_para(
+			100.into(),
+			Info {
+				scheduling: Scheduling::Always,
+			},
+			wasm.into(),
+			genesis_state.into(),
+		)
+		.into(),
+	));
+	let nonce = 0;
+	let period = BlockHashCount::get()
+		.checked_next_power_of_two()
+		.map(|c| c / 2)
+		.unwrap_or(2) as u64;
+	let tip = 0;
+	let extra: SignedExtra = (
+		OnlyStakingAndClaims,
+		frame_system::CheckVersion::<Runtime>::new(),
+		frame_system::CheckGenesis::<Runtime>::new(),
+		frame_system::CheckEra::<Runtime>::from(generic::Era::mortal(period, current_block)),
+		frame_system::CheckNonce::<Runtime>::from(nonce),
+		frame_system::CheckWeight::<Runtime>::new(),
+		pallet_transaction_payment::ChargeTransactionPayment::<Runtime>::from(tip),
+		registrar::LimitParathreadCommits::<Runtime>::new(),
+		parachains::ValidateDoubleVoteReports::<Runtime>::new(),
+	);
+	let raw_payload = SignedPayload::from_raw(
+		call.clone().into(),
+		extra.clone(),
+		(
+			(),
+			runtime_version.spec_version,
+			genesis_block,
+			current_block_hash,
+			(),
+			(),
+			(),
+			(),
+			(),
+		),
+	);
+	let signature = raw_payload.using_encoded(|e| Alice.sign(e));
+
+	// register parachain
+	let ex = polkadot_runtime::UncheckedExtrinsic::new_signed(
+		call.into(),
+		Alice.into(),
+		sp_runtime::MultiSignature::Sr25519(signature),
+		extra,
+	);
+	let _register_block_hash = async_std::task::block_on(async {
+		Author::submit_extrinsic(&mut client, format!("0x{}", hex::encode(ex.encode()))).await
+	})
+	.unwrap();
+
+	// run cumulus
+	let cumulus_dir = tempdir().unwrap();
+	let mut cumulus = Command::new(cargo_bin("cumulus-test-parachain-collator"))
+		.stdout(Stdio::piped())
+		.stderr(Stdio::piped())
+		.arg("--base-path")
+		.arg(cumulus_dir.path())
+		.arg("--")
+		.arg(format!(
+			"--bootnodes=/ip4/127.0.0.1/tcp/30333/p2p/{}",
+			polkadot_alice_id
+		))
+		.arg(format!(
+			"--bootnodes=/ip4/127.0.0.1/tcp/50666/p2p/{}",
+			polkadot_bob_id
+		))
+		.spawn()
+		.unwrap();
+	let mut cumulus_child = ChildHelper::new("cumulus", &mut cumulus);
+
+	// wait for blocks to be generated
+	thread::sleep(Duration::from_secs(60));
+
+	// check output
+	cumulus_child.terminate();
+	assert!(
+		cumulus_child
+			.read_stderr_to_end()
+			.unwrap()
+			.contains("best: #2"),
+		"no parachain blocks seems to have been produced",
+	);
```

Comment:
In the event that this test fails, it will be very useful to have the log output.

```suggestion
      let cumulus_log = cumulus_child.read_stderr_to_end().unwrap();
      println!("{}", cumulus_log);
	assert!(
		cumulus_log.contains("best: #2"),
		"no parachain blocks seems to have been produced",
	);
```

Reply:
It does already! Small summary: this whole helper struct has a String for stdout and stderr, every time you call a function to get the logs it first add to that string and then return a string slice to that string so nothing is never lost. When the helper is dropped, it always prints out stdout and stderr but if the test succeeds cargo doesn't show what has been printed.

Reply:
Is the `Drop` implementation guaranteed to fire before the failed assertion aborts the test? It may be worth writing a small secondary test which is `#[ignore]`d, just to demonstrate that that output does in fact get captured and emitted.

My impression was that `assert` macro failures are panics under the hood, and may not necessarily unwind all `Drop`s.

Reply:
assert does panic. panics call all the Drops. I don't see the value to adding such a test. The day this test fail someone will see the logs and if they don't fix it they will fix it. A test that needs to be run by human is guaranteed to be never run imo

Reply:
OK, fair enough.

Reply:
Wait, you said:
> may not necessarily unwind all Drops.

I read that it does unwind. Do you know for sure I should not rely on that behavior??

Reply:
I am not certain that you can not rely on that behavior; this isn't an area I've got a lot of experience in. I'm just aware of a bunch of chatter in the community that [you can't rely on `Drop` being executed in all cases](https://doc.rust-lang.org/nightly/nomicon/leaking.html).

Still, in this case at least, it looks like it does work properly: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=53f6cea04710cc1ed2fad0d34a18a3ee